src/software/new_geom/convex_polygon.cpp

```diff
@@ -167,3 +167,98 @@ double ConvexPolygon::area() const
 
     return std::abs(first_term - second_term) / 2;
 }
+
+Point ConvexPolygon::centroid() const
+{
+    // Explanation of the math/geometry behind this:
+    // https://fotino.me/calculating-centroids/
+    double x_centre    = 0;
+    double y_centre    = 0;
+    double signed_area = 0;
+
+    for (unsigned i = 0; i < points_.size(); i++)
+    {
+        double x0 = points_[i].x();
+        double y0 = points_[i].y();
+        double x1 = points_[(i + 1) % points_.size()].x();
+        double y1 = points_[(i + 1) % points_.size()].y();
+        double a  = (x0 * y1) - (x1 * y0);
+
+        x_centre += (x0 + x1) * a;
+        y_centre += (y0 + y1) * a;
+        signed_area += a;
+    }
+
+    return Point((Vector(x_centre, y_centre) / (3 * signed_area)));
+}
+
+ConvexPolygon ConvexPolygon::expand(const Vector& v) const
+{
+    // ASCII art showing an expanded ConvexPolygon
+    //
+    // Original ConvexPolygon:
+    //
+    //            B---------C
+    //            |         |
+    //            |         |
+    //            A---------D
+    //
+    //                 |
+    //                 V
+    //          expansion vector
+    //
+    // Expanded ConvexPolygon (A and D are shifted down):
+    //
+    //            B---------C
+    //            |         |
+    //            |         |
+    //            |         |
+    //            A'--------D'
+    //
+
+    std::vector<Point> expanded_points;
+    Point c = centroid();
+
+    enum Orientation
+    {
+        LEFT,
+        RIGHT,
+        ON_THE_LINE
+    };
+
+    // Draws a line perpendicular to v along c and returns which side of the line the
+    // point p is on
+    auto orientation = [&](const Point& p) {
+        Point b1 = c + v.perpendicular();
+        Point b2 = c - v.perpendicular();
+        double d =
+            (p.x() - b1.x()) * (b2.y() - b1.y()) - (p.y() - b1.y()) * (b2.x() - b1.x());
+        if (d == 0)
+        {
+            return Orientation::ON_THE_LINE;
+        }
+        else if (d < 0)
```

Comment:
I'm thinking LEFT and RIGHT might be confusing without context.

Is there anything else we could name these to make them more descriptive?

From what I see they are 'left' and 'right' of the direction of the Vector, but not necessarily the field?
@garethellis0 @MathewMacDougall

Reply:
I can see where there might have been some confusion here. Perhaps it's worth adding a comment clarifying that "left" and "right" are with respect to the vector pointing "straight up".

Reply:
updated the comments